### PR TITLE
Analysis proposal

### PR DESCRIPTION
**Pull request by hvanbakel**
So here is my proposal, there are a couple of things that I've done:
1. Created a number of interfaces that decouples the implementations from concrete types as that will allow us to be more flexible down the road and limit the risks of having to break the contract 
2. Split responsibilities, 
Analyzer knows how to take the diagnostics from the options and run them
Reporter knows how to report a diagnostic result without knowing anything about the project/analysis/conversion
Diagnostic knows only about a project and a diagnostic

These split responsibilities allow us to test the individual components more easily I've included an example test of one of the diagnostics. We can do the same for the analyzer and the reporter by mocking the dependencies. This will make the scope of the tests smaller (less code under test) resulting in more descriptive tests.

### CODE REVIEW

File: Project2015To2017/Analysis/AnalysisOptions.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/AnalysisOptions.cs:13]: Actually that could be used in future diagnostics.

File: Project2015To2017/Analysis/Analyzer.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/Analyzer.cs:39]: That's cool. It now actually takes care of each project and its own Solution property.

Comment (andrew-boyarshin) [reply Project2015To2017/Analysis/Analyzer.cs:39]: But now that I think of it, the removal of `RootDirectory` property from `AnalysisOptions` can cause code duplication. Oh well, maybe later add some extension method or utility to `DiagnosticBase`.

File: Project2015To2017/Analysis/DiagnosticBase.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/DiagnosticBase.cs:75]: The removal of this property makes me wonder why this was done. I agree with your static readonly approach rationale but still the assignment could be done in `DiagnosticBase` constructor. This way both Id & Code will be calculated once per instance (a bit worse than static+readonly, significantly better than my old solution), in one place, no extension method, no code duplication in every diagnostic. What do you think about taking this approach?

File: Project2015To2017/Analysis/DiagnosticBase.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/DiagnosticBase.cs]: This recalculation would be unnecessary if my new approach is taken.

File: Project2015To2017/Analysis/Diagnostics/W001IllegalProjectTypeDiagnostic.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/Diagnostics/W001IllegalProjectTypeDiagnostic.cs]: That's a tiny typo copy-pasted in other 2 places. :)

File: Project2015To2017/Analysis/Diagnostics/W001IllegalProjectTypeDiagnostic.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/Diagnostics/W001IllegalProjectTypeDiagnostic.cs:21]: Wow that's optimization with a huge assumption in mind. What if the diagnostic is expanded (realloc)? Why not leave capacity for List<T> to be decided? But yeah, that's minor point.

File: Project2015To2017/Analysis/Diagnostics/W020MicrosoftCSharpDiagnostic.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/Diagnostics/W020MicrosoftCSharpDiagnostic.cs]: Things seem inconsistent. I believe the namespace qualifier can be omitted in favor of `using` declaration.

File: Project2015To2017/Analysis/IReporterOptions.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/IReporterOptions.cs]: A bit of inconsistency here too. Namespace qualifier is better replaced with `using`.

File: Project2015To2017/Analysis/ReporterBase.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/ReporterBase.cs]: The original check is lost. If `result.Location.Source` is lost, `GetRelativePathTo` will throw `NullReferenceException`.

File: Project2015To2017/Analysis/ReporterOptions.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/ReporterOptions.cs]: Namespace again.

File: Project2015To2017/Analysis/ReporterOptions.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/ReporterOptions.cs:5]: Why `internal`? Wouldn't it make sense for this to be `public`, especially since it is `sealed` and concrete?

File: Project2015To2017Tests/W001IllegalProjectTypeDiagnosticTest.cs
Comment (andrew-boyarshin) [inline Project2015To2017Tests/W001IllegalProjectTypeDiagnosticTest.cs]: Some indentation issues.

File: Project2015To2017Tests/W001IllegalProjectTypeDiagnosticTest.cs
Comment (andrew-boyarshin) [inline Project2015To2017Tests/W001IllegalProjectTypeDiagnosticTest.cs]: Wow, that's a long line (and a couple more below). Why full qualification?

**Review by andrew-boyarshin — APPROVED** [review]
Well, I like your approach. I fetched the general idea from our previous discussion but this implementation exceeded my expectations (the devil is in the details, as always). I am about to merge this one as soon as I receive initial feedback on my comments. But the comments should still be addressed before merge into official repo. The main part is the Id & Code properties approach, but there are also indentation and codestyle changes to be done. I can do that in next commit on `analysis` branch after merge. Waiting for your reply.

File: Project2015To2017/Analysis/AnalysisOptions.cs
Comment (hvanbakel) [reply Project2015To2017/Analysis/AnalysisOptions.cs:13]: I would copy what is needed though to keep them separate concerns. Like what I did with reporteroptions. 
Including it like this means it needs to know everything about all the options, current and future. Maybe extract an interface and put that here? Having it coupled directly will complicate things in the long run

Comment (andrew-boyarshin) [reply Project2015To2017/Analysis/AnalysisOptions.cs:13]: All right, we will see what is needed when it will be needed and decide on the best approach then. :)

File: Project2015To2017/Analysis/DiagnosticBase.cs
Comment (hvanbakel) [reply Project2015To2017/Analysis/DiagnosticBase.cs:75]: I'm not too worried about either approach as long as we can write nice code opting out of a single one.

File: Project2015To2017/Analysis/DiagnosticBase.cs
Comment (hvanbakel) [reply Project2015To2017/Analysis/DiagnosticBase.cs]: Yes, go for it

File: Project2015To2017/Analysis/Diagnostics/W001IllegalProjectTypeDiagnostic.cs
Comment (hvanbakel) [reply Project2015To2017/Analysis/Diagnostics/W001IllegalProjectTypeDiagnostic.cs:21]: It only initializes more efficiently, it still will resize if you add more items than capacity

File: Project2015To2017/Analysis/Diagnostics/W020MicrosoftCSharpDiagnostic.cs
Comment (hvanbakel) [reply Project2015To2017/Analysis/Diagnostics/W020MicrosoftCSharpDiagnostic.cs]: Yes, this is just my vs settings, either way is fine

File: Project2015To2017/Analysis/Diagnostics/W001IllegalProjectTypeDiagnostic.cs
Comment (andrew-boyarshin) [reply Project2015To2017/Analysis/Diagnostics/W001IllegalProjectTypeDiagnostic.cs:21]: Yeah, I now it can still resize (with realloc), but the microoptimization is cute. :)

File: Project2015To2017/Analysis/ReporterBase.cs
Comment (hvanbakel) [reply Project2015To2017/Analysis/ReporterBase.cs]: Right, my assumption is that location will be null if you don't have a file

File: Project2015To2017/Analysis/ReporterOptions.cs
Comment (hvanbakel) [reply Project2015To2017/Analysis/ReporterOptions.cs:5]: I think it is better to keep the interface as the public API not the types

File: Project2015To2017/Analysis/Analyzer.cs
Comment (hvanbakel) [reply Project2015To2017/Analysis/Analyzer.cs:39]: I think duplicating trivial code is necessary when it keeps components separate. Like you said, there are other ways to solve this like extension methods.

File: Project2015To2017/Analysis/ReporterOptions.cs
Comment (andrew-boyarshin) [reply Project2015To2017/Analysis/ReporterOptions.cs:5]: Let's imagine one needs to implement his own Reporter with his its own option set. We always pass new `ReporterOptions` instance in `Analyzer`. There is actually no way for user to specify his own custom options. Feels like this point is missing in `Analyzer`.

Comment (hvanbakel) [reply Project2015To2017/Analysis/ReporterOptions.cs:5]: Correct, extensibility has not gotten the thought needed yet. The problem is that assignment of root directory. We need to think about how to mitigate that

**Comment by andrew-boyarshin** [comment]
I intend to revert a part of it related to constructors removal, and implement my new approach to Id & Code. This I will add as a commit to this PR, then next commit would be indentation and namespaces fix, then merge to `analysis`. After that I will play around with code and see how best to tackle `RootDirectory` issue. Then I request a new review on the original PR in official repo.

**Comment by hvanbakel** [comment]
Sounds good, I think the solution might be to just include the project file and directory and move the logic to calculate the root folder into the reporter. That would mean there's no need to extend reporter Options

File: Project2015To2017/Analysis/ExtensionMethods.cs
Comment (andrew-boyarshin) [inline Project2015To2017/Analysis/ExtensionMethods.cs]: I decided to keep this one to later move `Transforms.Helpers.GetRelativePathTo` to this class.

**Pull request merged**
